builder/builder-next/builder.go

```diff
@@ -511,17 +515,32 @@ func toBuildkitExtraHosts(inp []string) (string, error) {
 }
 
 func toBuildkitPruneInfo(opts types.BuildCachePruneOptions) (client.PruneInfo, error) {
-	var unusedFor time.Duration
-	unusedForValues := opts.Filters.Get("unused-for")
+	var until time.Duration
+	untilValues := opts.Filters.Get("until")
+	unusedForValues := opts.Filters.Get("unused-for") // synonym for until
 
-	switch len(unusedForValues) {
+	switch len(untilValues) {
```

Comment:
I find this switch a bit hard to grasp, as it's doing two things (validating, and handling of the values).

How about;

```go
func toBuildkitPruneInfo(opts types.BuildCachePruneOptions) (client.PruneInfo, error) {
	var until time.Duration
	untilValues := opts.Filters.Get("until")
	unusedForValues := opts.Filters.Get("unused-for") // synonym for until

	if len(untilValues) > 0 && len(unusedForValues) > 0 {
		return client.PruneInfo{}, errSynonymFilterValues
	}
	untilValues = append(untilValues, unusedForValues...)

	switch len(untilValues) {
	case 0:
		// nothing to do
	case 1:
		var err error
		until, err = time.ParseDuration(untilValues[0])
		if err != nil {
			return client.PruneInfo{}, errors.Wrap(err, "until filter expects a duration (e.g., '24h')")
		}

	default:
		return client.PruneInfo{}, errMultipleFilterValues
	}

	bkFilter := make([]string, 0, opts.Filters.Len())
	for cacheField := range cacheFields {
		if opts.Filters.Include(cacheField) {
			values := opts.Filters.Get(cacheField)
			switch len(values) {
			case 0:
				bkFilter = append(bkFilter, cacheField)
			case 1:
				if cacheField == "id" {
					bkFilter = append(bkFilter, cacheField+"~="+values[0])
				} else {
					bkFilter = append(bkFilter, cacheField+"=="+values[0])
				}
			default:
				return client.PruneInfo{}, errMultipleFilterValues
			}
		}
	}
	return client.PruneInfo{
		All:          opts.All,
		KeepDuration: until,
		KeepBytes:    opts.KeepStorage,
		Filter:       []string{strings.Join(bkFilter, ",")},
	}, nil
}
```

```patch
diff --git a/builder/builder-next/builder.go b/builder/builder-next/builder.go
index f682b88e8f..1bd2a11be1 100644
--- a/builder/builder-next/builder.go
+++ b/builder/builder-next/builder.go
@@ -519,24 +519,15 @@ func toBuildkitPruneInfo(opts types.BuildCachePruneOptions) (client.PruneInfo, e
        untilValues := opts.Filters.Get("until")
        unusedForValues := opts.Filters.Get("unused-for") // synonym for until
 
+       if len(untilValues) > 0 && len(unusedForValues) > 0 {
+               return client.PruneInfo{}, errSynonymFilterValues
+       }
+       untilValues = append(untilValues, unusedForValues...)
+
        switch len(untilValues) {
        case 0:
-               switch len(unusedForValues) {
-               case 0:
-
-               case 1:
-                       var err error
-                       until, err = time.ParseDuration(unusedForValues[0])
-                       if err != nil {
-                               return client.PruneInfo{}, errors.Wrap(err, "unused-for filter expects a duration (e.g., '24h')")
-                       }
-               default:
-                       return client.PruneInfo{}, errMultipleFilterValues
-               }
+               // nothing to do
        case 1:
-               if len(unusedForValues) > 0 {
-                       return client.PruneInfo{}, errSynonymFilterValues
-               }
                var err error
                until, err = time.ParseDuration(untilValues[0])
                if err != nil {
```

Reply:
Could actually better use the`Filter.Contains(..` for that;

```go
if opts.Filters.Contains("until") && opts.Filters.Contains("unused-for") {
	return client.PruneInfo{}, errSynonymFilterValues
}
```

```patch
diff --git a/builder/builder-next/builder.go b/builder/builder-next/builder.go
index f682b88e8f..a2dd035e06 100644
--- a/builder/builder-next/builder.go
+++ b/builder/builder-next/builder.go
@@ -515,28 +515,16 @@ func toBuildkitExtraHosts(inp []string) (string, error) {
 }
 
 func toBuildkitPruneInfo(opts types.BuildCachePruneOptions) (client.PruneInfo, error) {
+       if opts.Filters.Contains("until") && opts.Filters.Contains("until") {
+               return client.PruneInfo{}, errSynonymFilterValues
+       }
+
        var until time.Duration
-       untilValues := opts.Filters.Get("until")
-       unusedForValues := opts.Filters.Get("unused-for") // synonym for until
+       untilValues := append(opts.Filters.Get("until"),  opts.Filters.Get("unused-for")...)
 
        switch len(untilValues) {
        case 0:
-               switch len(unusedForValues) {
-               case 0:
-
-               case 1:
-                       var err error
-                       until, err = time.ParseDuration(unusedForValues[0])
-                       if err != nil {
-                               return client.PruneInfo{}, errors.Wrap(err, "unused-for filter expects a duration (e.g., '24h')")
-                       }
-               default:
-                       return client.PruneInfo{}, errMultipleFilterValues
-               }
        case 1:
-               if len(unusedForValues) > 0 {
-                       return client.PruneInfo{}, errSynonymFilterValues
-               }
                var err error
                until, err = time.ParseDuration(untilValues[0])
                if err != nil {
```